### PR TITLE
support for virtual goods

### PR DESCRIPTION
**Pull request by danac-gs**

### CODE REVIEW

File: Helper/Cart.php [hunk 1]
```diff
@@ -474,21 +474,21 @@ public function getCartData($payment_only, $place_order_payload, $quote = null)
             $place_order_payload = json_decode($place_order_payload);
 
             $email                = @$place_order_payload->email ?: $email;
-            $billingAddress       = @$place_order_payload->billingAddress;
-            $billingStreetAddress = (array)@$billingAddress->street;
+            $billing_address      = @$place_order_payload->billingAddress;
```
Comment (daisy1754) [inline Helper/Cart.php]: nit: let's stick to camelCase (billingAddress) per magento's convention.

File: Helper/Cart.php [hunk 2]
```diff
@@ -497,58 +497,67 @@ public function getCartData($payment_only, $place_order_payload, $quote = null)
...
+                        unset($shipping_address);
+                        break;
+                    }
                 }
-            }
 
-            if (@$shipping_address) {
-                $cost         = $shippingAddress->getShippingAmount();
-                $rounded_cost = $this->getRoundAmount($cost);
+                if (@$shipping_address) {
```
Comment (daisy1754) [inline Helper/Cart.php]: looks like this `$shipping_address` is always defined when code executes here?

File: Helper/Cart.php [hunk 3]
```diff
@@ -497,58 +497,67 @@ public function getCartData($payment_only, $place_order_payload, $quote = null)
...
-                ]];
+                    $cart['shipments'] = [[
+                        'cost' => $rounded_cost,
+                        'tax_amount' => $this->getRoundAmount($address->getShippingTaxAmount()),
+                        'shipping_address' => $shipping_address,
+                        'service' => $shippingAddress->getShippingDescription(),
+                        'reference' => $shippingAddress->getShippingMethod(),
+                    ]];
+                }
+            } else {
```
Comment (daisy1754) [inline Helper/Cart.php]: nit: `if not A else B` is less readable. Let's change this to `if virtual then B else A`

File: Helper/Order.php [hunk 4]
```diff
@@ -549,17 +549,13 @@ private function deleteRedundantQuotes($quote) {
...
 
-        // load the quote from reserved order id
-        // $quote = $this->loadQuote($incrementId);
-
         // Load quote from entity id
         $quoteId = $transaction->order->cart->order_reference;
-        $quote = $this->quoteFactory->create()->load($quoteId);
-        //$quote = $this->quoteRepository->get($quoteId);
+        //$quote = $this->quoteFactory->create()->load($quoteId);
+        $quote = $this->quoteRepository->get($quoteId);
```
Comment (daisy1754) [inline Helper/Order.php]: what is the reason for this change?

File: Model/Api/ShippingMethods.php [hunk 5]
```diff
@@ -341,12 +341,12 @@ public function getShippingMethods($cart, $shipping_address)
      * Get Shipping and Tax from cache or run the Shipping options collection routine, store it in cache and return.
      *
      * @param Quote $quote
-     * @param array $shipping_address
+     * @param array $address_data
```
Comment (daisy1754) [inline Model/Api/ShippingMethods.php]: nit: let's use camelCase - addressData

File: Model/Api/ShippingMethods.php [hunk 6]
```diff
@@ -378,49 +378,47 @@ function ($el) {
...
+            'country_id' => @$address_data['country_code'],
+            'postcode'   => @$address_data['postal_code'],
+            'region'     => @$address_data['region'],
             'region_id'  => $region ? $region->getId() : null,
-            'firstname'  => @$shipping_address['first_name'],
-            'lastname'   => @$shipping_address['last_name'],
-            'street'     => trim(@$shipping_address['street_address1'] . "\n" . @$shipping_address['street_address2']),
-            'city'       => @$shipping_address['locality'],
-            'telephone'  => @$shipping_address['phone'],
+            'street'     => trim(@$address_data['street_address1'] . "\n" . @$address_data['street_address2']),
```
Comment (daisy1754) [inline Model/Api/ShippingMethods.php]: I think city may be needed for shipping calculation?

File: Model/Api/ShippingMethods.php [hunk 7]
```diff
@@ -441,22 +439,43 @@ function ($el) {
...
+
+            $this->totalsCollector->collectAddressTotals($quote, $billingAddress);
+            $tax_amount = $this->cartHelper->getRoundAmount($billingAddress->getTaxAmount());
+
+            return [
+                $this->shippingOptionInterfaceFactory
+                    ->create()
+                    ->setService('No Shipping Required')
+                    ->setCost(0)
+                    ->setReference(null)
```
Comment (daisy1754) [inline Model/Api/ShippingMethods.php]: I suggest you define explicit magic constant string as opposed to using null.

File: Model/Api/ShippingMethods.php [hunk 8]
```diff
@@ -441,22 +439,43 @@ function ($el) {
...
      *
      * @return ShippingOptionInterface[]
      */
-    private function getShippingOptions($quote, $shipping_address)
+    private function getShippingOptions($quote, $address_data)
     {
+        if ($quote->isVirtual()) {
+
+            $billingAddress = $quote->getBillingAddress();
+            $billingAddress->addData($address_data);//->save();
```
Comment (daisy1754) [inline Model/Api/ShippingMethods.php]: nit: no comment out code pls

File: Setup/UpgradeSchema.php [hunk 9]
```diff
@@ -27,10 +27,16 @@ public function upgrade(SchemaSetupInterface $setup, ModuleContextInterface $con
                 'nullable' => true,
                 'default' => null,
                 'unsigned' => true,
-                'comment' => 'Original Quote ID',
+                'comment' => 'Original Quote ID'
             ]
         );
 
+        $setup->getConnection()->addIndex(
```
Comment (daisy1754) [inline Setup/UpgradeSchema.php]: please do this in separate PR

File: view/frontend/templates/js/replacejs.phtml [hunk 10]
```diff
@@ -153,590 +150,687 @@ window.boltConfig = <?php echo $block->getSettings(); ?>;
...
+            if (create_request) create_request.abort();
+            // preset success condition to false
+            cart.orderToken = '';
+            // define the params sent with the request variable
+            var params = [];
+
+            // check and set payment_only flag
+            if (getCheckoutType() === 'payment') {
+
+                // stop if customer email field is present and invalid
```
Comment (daisy1754) [inline view/frontend/templates/js/replacejs.phtml]: not sure if changes in this file is relevant for virtual items upport

File: Helper/Cart.php (hunk 1 above)
Comment (danac-gs) [reply Helper/Cart.php]: done for the whole file

File: Helper/Cart.php (hunk 2 above)
Comment (danac-gs) [reply Helper/Cart.php]: it is conditionally unset a few lines above

File: Helper/Cart.php (hunk 3 above)
Comment (danac-gs) [reply Helper/Cart.php]: done

File: Helper/Order.php (hunk 4 above)
Comment (danac-gs) [reply Helper/Order.php]: Model::load is deprecated (but still heavily used). The repository has already been available. Removed the factory completely.

File: Model/Api/ShippingMethods.php (hunk 5 above)
Comment (danac-gs) [reply Model/Api/ShippingMethods.php]: done for the file

File: Model/Api/ShippingMethods.php (hunk 6 above)
Comment (danac-gs) [reply Model/Api/ShippingMethods.php]: It is not needed, but it was there. I probably deleted it by mistake. It's back.

File: Model/Api/ShippingMethods.php (hunk 7 above)
Comment (danac-gs) [reply Model/Api/ShippingMethods.php]: done

File: Model/Api/ShippingMethods.php (hunk 8 above)
Comment (danac-gs) [reply Model/Api/ShippingMethods.php]: ok

File: view/frontend/templates/js/replacejs.phtml (hunk 10 above)
Comment (danac-gs) [reply view/frontend/templates/js/replacejs.phtml]: More than relevant. The file is pretty much refactored, some Magento objects are required. There is a *is guest checkout allowed* validation / login popup on bolt button click. Virtual / downloadable goods are allowed to logged in customers only on most setups.

**Comment by danac-gs** [comment]
@daisy1754 please do not merge yet. I have to double check all the changes.

File: Helper/Order.php [hunk 11]
```diff
@@ -320,10 +311,12 @@ public function fetchTransactionInfo($reference)
      */
     private function setShippingMethod($quote, $transaction)
     {
+        if ($quote->isVirtual()) return;
+
         $shippingAddress = $quote->getShippingAddress();
         $shippingAddress->setCollectShippingRates(true);
 
-        $shippingMethod = $transaction->order->cart->shipments[0]->reference;
+        $shippingMethod = @$transaction->order->cart->shipments[0]->reference ?: null;
```
Comment (daisy1754) [inline Helper/Order.php]: I guess now you get NO_SHIPPING_REFERENCE here so it never become null?

File: Helper/Order.php [hunk 12]
```diff
@@ -549,17 +542,12 @@ private function deleteRedundantQuotes($quote) {
...
         $incrementId = $transaction->order->cart->display_id;
 
-        // load the quote from reserved order id
-        // $quote = $this->loadQuote($incrementId);
-
         // Load quote from entity id
         $quoteId = $transaction->order->cart->order_reference;
-        $quote = $this->quoteFactory->create()->load($quoteId);
-        //$quote = $this->quoteRepository->get($quoteId);
+        $quote = $this->quoteRepository->get($quoteId);
```
Comment (YevhenShBolt) [inline Helper/Order.php]: Can you move it to the separate method (like "getQuoteById()")
And then we can cover it with PHPUnit test easily.

Comment (danac-gs) [reply Helper/Order.php]: Quote loading and saving moved to cart helper
